tutorials/haskell/generics/tutorial.md

```diff
@@ -0,0 +1,584 @@
+
+* If no `K1` instance is given, the function may still work for enumeration
+  types, where no constructor has any fields.
+
+* If no `V1` instance is given, the function may still work for any datatype
+  that is not empty.
+
+* If no `U1` instance is given, the function may still work for any datatype
+  where each constructor has at least one field.
+
+An `M1` instance is always required (but it can just ignore the
+meta-information).
+
+## Putting the magic into the type classes
+
+Having dealt with generic implementation of the functionality of interest,
+let's put it all together and use a special GHC extension to allow the user
+derive type classes without knowing anything about generics.
+
+For generic implementation to work without user's definition we need to
+provide it as the default definition. As you have already seen, generic
+implementation often involves `Generic` constraints and it would be ugly to
+add them to every type class (on the left hand side of `=>`) just to make
+deriving easier. The `default` keyword, enabled by the `DefaultSignartures`
+language extension, allows to give a different type signature for default
+implementation of a type class:
+
+```haskell
+instance Functor f where
+  fmap :: (a -> b) -> f a -> f b
+  default fmap :: (Generic1 f, Functor (Rep1 f)) => (a -> b) -> f a -> f b
+  fmap = to1 . fmap . from1
+
+class CountFields a where
+  countFields :: a -> Natural
+  default countFields :: (Generic a, CountFields1 (Rep a)) => a -> Natural
+  countFields = defaultCountFields
+```
+
+This way we can eat the cake and have it: deriving is easy and otherwise no
+ugly details is visible!
+
+## Conclusion
+
+Generics is a powerful means to automate error-prone and boring deriving.
+The feature is however helpful beyond just that use-case, as with a bit of
+creativity it allows to reason about data types generically (for example, we
+could calculate number of constructors at compile time), and generate values
+in a type safe way (useful, for example, in combination with Quick Check's
+value generation). Finally, there are quite a few very interesting packages
+that complement or build on top of GHC generics. Once you feel comfortable
+with vanilla generics,
+libraries [`generic-sop`](https://hackage.haskell.org/package/generics-sop)
+may be of interest.
+
+## See also
+
+A set of recommended resources on the topic:
+
+* [A generic deriving Mechanism for Haskell (paper)](http://dreixel.net/research/pdf/gdmh.pdf)
+* [Documentation for `GHC.Generics` on Hackage](https://hackage.haskell.org/package/base-4.8.2.0/docs/GHC-Generics.html)
+* [The page on Haskell Wiki](https://wiki.haskell.org/GHC.Generics)
+* [The `generics-sop` package on Hackage](https://hackage.haskell.org/package/generics-sop)
+* [Typeable and Data in Haskell](http://chrisdone.com/posts/data-typeable) —
+  the blog by Chris Done contains some straightforward examples of generic
+  programming using `Typeable` and `Data` type class
```

Comment:
Please add at the end a you might also like to read list of blogs or tutorials you thing the audience of this tutorial might like from our own site, we'd like to give people an outlook on other interesting things we've been working on.